lib/Parse/ParseGeneric.cpp

```diff
@@ -350,9 +350,23 @@ ParserStatus Parser::parseGenericWhereClause(
         SecondType = makeParserResult(new (Context) ErrorTypeRepr(PreviousLoc));
 
       // Add the requirement
-      Requirements.push_back(RequirementRepr::getSameType(FirstType.get(),
-                                                      EqualLoc,
-                                                      SecondType.get()));
+      if (FirstType.hasCodeCompletion()) {
+        // If the first type has a code completion token, don't record a same
+        // type constraint because otherwise if we have
+        //   K.#^COMPLETE^# == Foo
+        // we parse this as
+        //   K == Foo
+        // and thus simplify K to Foo. But we didn't want to state that K is Foo
+        // but that K has a member of type Foo.
+        // FIXME: The proper way to fix this would be to represent the code
+        // completion token in the TypeRepr.
+        Requirements.push_back(RequirementRepr::getTypeConstraint(
+            FirstType.get(), PreviousLoc,
+            new (Context) ErrorTypeRepr(PreviousLoc)));
```

Comment:
It doesn't really matter, but I don't think we need to use fake locations.
```suggestion
            FirstType.get(), EqualLoc,
            new (Context) ErrorTypeRepr(SecondType.getLoc())));
```

Reply:
Yes, you’re right. That’s what I got from copy-pasting the fake requirement from below.